Issue #5648: Create UTs to control xpath-suppression model regression

timurt opened the pull request:
https://github.com/checkstyle/checkstyle/issues/5648

rnveach commented:
@timurt CI is failing.

rnveach commented:
@timurt CI is still failing.

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/XpathRegressionTest.java:
```diff
@@ -0,0 +1,199 @@
+        verify(checkerConfig,
+                file.getPath(),
+                expectedViolation);
+    }
```
is this method required ?

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/XpathRegressionTest.java:
```diff
@@ -0,0 +1,199 @@
+
+        final String[] expectedViolation = {
+            "4:5: " + getCheckMessage(CyclomaticComplexityCheck.class,
+                    CyclomaticComplexityCheck.MSG_KEY, 2, 0),
```
```
+        final int violationLineNumber = 4;
+        final int violationColumnNumber = 5;
..
+            "4:5: " + getCheckMessage
```

unfortunate copy-paste, can we eliminate it ?

Why you limit design only to one violation in file ?

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionxpathfilter/InputSuppressionXpathFilterCyclomatic.java:
```diff
@@ -0,0 +1,11 @@
+package com.puppycrawl.tools.checkstyle.filters.suppressionxpathfilter;
+
+public class InputSuppressionXpathFilterCyclomatic {
```
it might be related to my above question about single violation in file....

if we go this way, we will definitely need several test inputs for each check, to test its ability to suppress different types of violations(on different AST elements), so we need to create extra folder for each check under "suppressionxpathfilter".

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/XpathRegressionTest.java:
```diff
@@ -0,0 +1,199 @@
+            bw.write(checkName);
+            bw.write("\"\n");
+            bw.write("       query=\"");
+            bw.write(xpathQueries.stream().collect(Collectors.joining(DELIMITER)));
```
why we join all xpath queries ?

Is it better to test each query separately ? to make sure that all that we generated could suppress violation.

MEZk replied:
Yes, we need to test generated queries one by one.

romani replied:
@timurt  , what is status on this point? Is there any good reason?
Please keep replying to each point, it help a lot.

romani commented:
@timurt , please find time to finish this PR, let us know if you have problems .

romani commented:
@timurt , ping. Please finish this PR.

rnveach commented:
@romani I still think we should consider some type of more automated approach to testing xpath then building all these tests manually.

romani commented:
We have not power to finish even this, more automated approach sounds like not achivable at all.

rnveach commented:
@romani I created a working automated UT regression at https://github.com/rnveach/checkstyle/commits/issue_5648 . Any issues it found I created issues for already. The logic of the UT is talked about at https://github.com/checkstyle/checkstyle/issues/5648#issuecomment-384612828 .

romani commented:
ok, lets start PR and pass it through process.

rnveach commented:
There's too many issues right now to put it to PR. https://github.com/checkstyle/checkstyle/issues/5821 or something similar causes alot of files to fail on either DOT or METHOD_CALL tokens. It is also a long test and I haven't even gotten it through half of our input files and it is taking 285 seconds. This is because it re-processes the file for every token it finds.

timurt commented:
@romani 
Sorry for delay, I fixed points you mentioned
Now I extract line and column number information from violation messages, so one file can have multiple violations

Also I found one bug with xpath queries, thats why I added [this sorting](https://github.com/timurt/checkstyle/commit/f181b69dffef085d7d443097ba0ea96df6a75917#diff-ad589c3bb03cfde0eefbf5df1782e29eR80)
It seems like the order of subqueries matters
Example
First query
```
/CLASS_DEF[@text='InputSuppressionXpathFilterCyclomatic']/OBJBLOCK/METHOD_DEF[@text='test'] | 
                    /CLASS_DEF[@text='InputSuppressionXpathFilterCyclomatic']/OBJBLOCK/METHOD_DEF[@text='test']/MODIFIERS | 
                    /CLASS_DEF[@text='InputSuppressionXpathFilterCyclomatic']/OBJBLOCK/METHOD_DEF[@text='foo2']
```

Second query
```
/CLASS_DEF[@text='InputSuppressionXpathFilterCyclomatic']/OBJBLOCK/METHOD_DEF[@text='test'] | 
                    /CLASS_DEF[@text='InputSuppressionXpathFilterCyclomatic']/OBJBLOCK/METHOD_DEF[@text='foo2'] |
                    /CLASS_DEF[@text='InputSuppressionXpathFilterCyclomatic']/OBJBLOCK/METHOD_DEF[@text='test']/MODIFIERS
```
First query returns two elements, the second one three elements

rnveach commented:
@timurt If you have found a problem separate from this issue, can you create a new issue for it?

romani commented:
@timurt , please be NOT stopped by any new bug you found, keep sorting with comment "till ...." , Please address my points and let's merge this PR.

romani commented:
>  It is also a long test and I haven't even gotten it through half of our input files and it is taking 285 seconds(~5min)

@rnveach ,
it will make our ITs to run too long, it might be better to move such test in separate project and just just reuse it in CI. Or may be to keep in in main project by run this test only by special profile activation.

>  I fixed points you mentioned

@timurt , ping.
You fixed only one point with copy-paste. What about removal of join of xpath and apply them separately ? Is there any additional problems, can we postpone them ?

timurt commented:
@romani 
Made two separate tests for each check
Made regression tests for 6 checks

If it is ok, I will finish other checks today

romani commented on src/test/java/com/puppycrawl/tools/checkstyle/XpathRegressionTest.java:
```diff
@@ -0,0 +1,505 @@
+    }
+
+    private String getPath(String subpackage, String filename) throws IOException {
+        return new File("src/test/resources/" + getPackageLocation()
```
should be `src/it/resources`

timurt replied:
done

romani commented on src/test/resources/com/puppycrawl/tools/checkstyle/filters/suppressionxpathfilter/cyclomaticcomplexitycheck/InputSuppressionXpathFilterCyclomaticTwo.java:
```diff
@@ -0,0 +1,25 @@
+public class InputSuppressionXpathFilterCyclomaticTwo {
+
+
+    public void foo2() {
```
as we expect only one violation in file, please put trailing comment where violation is expected.
"// violation".
Please do this for all input files.

timurt replied:
I used `//warn`, because [AbstractModuleTestSupport](https://github.com/checkstyle/checkstyle/blob/master/src/it/java/com/google/checkstyle/test/base/AbstractModuleTestSupport.java#L78) requires this comment

timurt commented on src/test/java/com/puppycrawl/tools/checkstyle/XpathRegressionTest.java:
```diff
@@ -0,0 +1,505 @@
+////////////////////////////////////////////////////////////////////////////////
```
> base on #5648 (comment)

probably I missed this comment from you
Do you want me to make separate test classes for each check or it is ok to have one big test file?

romani replied:
@timurt ,

after conversation with @rnveach : 
please just move this Test to `it` folder, as one test class.
In future .... ones we end up with this file to be huge we can refactor it to multiple test classes grouped by Check package number, example `XpathRegressionCodingTest.java`.

Expected path:
`checkstyle/src/it/java/org/checkstyle/suppressionxpathfilter/XpathRegressionTest.java`

`org/checkstyle` because own this domain, may be one day in far future we will migrate to new package names.

timurt replied:
done

romani commented:
@timurt ,
travis is not haypy again 

rename methodFallThruCCustomWords ==> methodFallThruCustomWords

romani commented on src/it/java/com/google/checkstyle/test/base/AbstractModuleTestSupport.java:
```diff
@@ -199,8 +204,10 @@ protected static DefaultConfiguration createTreeWalkerConfig(Configuration confi
-        final DefaultConfiguration dc = new DefaultConfiguration("root");
-        dc.addChild(config);
+        final DefaultConfiguration dc = new DefaultConfiguration(ROOT_MODULE_NAME);
+        if (config != null) {
```
why check for null become required, you do not pass null at AbstractXpathRegressionTest

timurt replied:
Yeah, I can revert it, it was one of my tries to fix errors

Btw, I took it from `AbstractModuleTestSupport` located in `test` package, should I revert it?

timurt replied:
Done

romani commented on src/it/java/com/google/checkstyle/test/base/AbstractModuleTestSupport.java:
```diff
@@ -155,11 +157,14 @@ public final Checker createChecker(Configuration moduleConfig)
         }
+        else if (ROOT_MODULE_NAME.equals(moduleConfig.getName())) {
+            dc = moduleConfig;
+        }
```
please explain reason of this update , following `else` create RootConfig.

timurt replied:
Because, otherwise it fails with 'Unable to instantiate "root" class, it is also not possible to instantiate it as "rootCheck" ...'

The problem, when you call `createRootConfig()` method, it creates configuration with name root (`new DefaultConfiguration("root")`).  And then you call `createChecker(Configuration moduleConfig, ModuleCreationOption moduleCreationOption)`, it goes to `else` section and creates `root` configuration with `root` child. So when this configuration is being passed to `Checker` object, it fails, because there is no such module called `root`.

I took this `else if` part from original `AbstractModuleTestSupport`

romani replied:
ok, we will think about this in next PR.

romani commented on src/it/java/org/checkstyle/suppressionxpathfilter/AbstractXpathRegressionTest.java:
```diff
@@ -0,0 +1,207 @@
+import org.junit.Rule;
+import org.junit.rules.TemporaryFolder;
+
+import com.google.checkstyle.test.base.AbstractModuleTestSupport;
```
it is unfortunate dependency between ITs, but we will think about it later.

timurt replied:
should I create new `AbstractModuleTestSupport`?

romani replied:
yes, but lets do this as next PR. We need to merge this PR.

romani commented on src/it/java/org/checkstyle/suppressionxpathfilter/AbstractXpathRegressionTest.java:
```diff
@@ -0,0 +1,207 @@
+            bw.write(checkName);
+            bw.write("\"\n");
+            bw.write("       query=\"");
+            bw.write(xpathQueries.stream().collect(Collectors.joining(DELIMITER)));
```
why "join" ? I asked this questions before.
if you postponing it for another issue please reference it or define reason.

timurt replied:
I probably misunderstood you
There are two ways
1. Create many `SuppressionXpathFilter` modules with only xpath query
2. Create one `SuppressionXpathFilter` with joined xpath query

I see no difference between them, the only difference in the second case there is an additional opportunity to test complex, combined queries

romani replied:
Why we need all of queries to suppress violation but not the one one of them ?

timurt replied:
Because all these queries point to the same line and column number

How do we know which query is responsible for node we want to suppress, since violation message does not contain token type information, only line and column?

So for each test we can manually specify required xpath query, or use one of two approaches I described above

timurt replied:
> my experiment:

First of all about different column numbers
The reason of this difference, that user messages consider tabs and start with 1, when `AstTreeStringPrinter ` prints 0-based column numbers without calculating tabs.

So If you want numbers to be the same, we should pass tabwidth to `AstTreeStringPrinter` and little change one line and thats all

> So if we force Checks to report issue on AST nodes that have text from file

You are suggesting to change `log()` calls inside checks, so Checks will report violation on nearest node that satisfies condition `node.getText()` not equals string representation of the `node.getType()`.
if you mean it, it will require a lot of correcting tests. But yes, probably it will work

Another way is to print `tokenType` inside violation messages, for instance
`InputTest.java:2:5:METHOD_DEF: Cyclomatic Complexity is 2 (max allowed is 0). [CyclomaticComplexity]`
And make xpath query generator accept tokenType among line and column numbers

Third approach is very similar to [#102](https://github.com/checkstyle/checkstyle/issues/102)
This is more natural for me and requires much less pain to make it
We could force users to use configuration file alongside with query generator parameter
So if you want to generate xpath query will use the following format
`java -jar checkstyle-8.10-all.jar -c checks.xml -s 4:5 TestClass.java`
Since we pass configuration file, we will be able to extract `LocalizedMessage` from errors list during runtime and generate xpath queries

romani replied:
Lets move out of this huge PR:
#5877 - AstTreeStringPrinter should use 1-based column and consider tab size
#5878 - Generate single xpath query by CLI

timurt commented:
@romani 

> rename methodFallThruCCustomWords ==> methodFallThruCustomWords

Why then [this](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/fallthrough/InputFallThrough.java#L363) and [this](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/nofinalizer/InputNoFinalizerFallThrough.java#L363) does not fail?

romani commented:
> Why then this and this does not fail?

hard to say, please satisfy `./.ci/test-spelling-unknown-words.sh` on your local.

romani commented:
@rnveach , please do final review.
I am already blind a bit, if you find any minors, it is ok, lets postpone them for later, lets merge this huge and long PR.

rnveach commented on src/test/java/com/puppycrawl/tools/checkstyle/internal/AllTestsTest.java:
```diff
@@ -209,7 +209,8 @@ private static void verifyHasProductionFile(Map<String, List<String>> allTests,
                     // tests external utility XPathEvaluator
-                    && !"XpathMapper.java".equals(fileName)) {
+                    && !"XpathMapper.java".equals(fileName)
+                    && !"XpathRegression.java".equals(fileName)) {
```
These can be removed.
